haystack/pipelines/base.py

```diff
@@ -498,67 +353,21 @@ def add_node(self, component: BaseComponent, name: str, inputs: List[str]):
         self._set_sub_component_names(component, component_names=component_names)
-        self.graph.add_node(name, component=component, inputs=inputs)
 
-        if len(self.graph.nodes) == 2:  # first node added; connect with Root
-            if not len(inputs) == 1 and inputs[0].split(".")[0] == self.root_node:
-                raise PipelineConfigError(
-                    f"The '{name}' node can only input from {self.root_node}. "
-                    f"Set the 'inputs' parameter to ['{self.root_node}']"
-                )
-            # edges_count = self.graph.graph.
-            self.graph.add_edge(self.root_node, name, label="output_1")
-            return
-
-        for input_node in inputs:
-            if "." in input_node:
-                [input_node_name, input_edge_name] = input_node.split(".")
-                if not "output_" in input_edge_name:
-                    raise PipelineConfigError(f"'{input_edge_name}' is not a valid edge name.")
-
-                outgoing_edges_input_node = self.graph.nodes[input_node_name]["component"].outgoing_edges
-                if not int(input_edge_name.split("_")[1]) <= outgoing_edges_input_node:
-                    raise PipelineConfigError(
-                        f"Cannot connect '{input_edge_name}' from '{input_node_name}' as it only has "
-                        f"{outgoing_edges_input_node} outgoing edge(s)."
-                    )
-            else:
-                try:
-                    outgoing_edges_input_node = self.graph.nodes[input_node]["component"].outgoing_edges
-                    if not outgoing_edges_input_node == 1:
-                        raise PipelineConfigError(
-                            f"Adding an edge from {input_node} to {name} is ambiguous as {input_node} has {outgoing_edges_input_node} edges. "
-                            f"Please specify the output explicitly."
-                        )
-
-                except KeyError as e:
-                    raise PipelineConfigError(
-                        f"Cannot find node '{input_node}'. Make sure you're not using more "
-                        f"than one root node ({valid_root_nodes}) in the same pipeline and that a node "
-                        f"called '{input_node}' is defined."
-                    ) from e
-
-                input_node_name = input_node
-                input_edge_name = "output_1"
-            self.graph.add_edge(input_node_name, name, label=input_edge_name)
-
-        if not nx.is_directed_acyclic_graph(self.graph):
-            self.graph.remove_node(name)
-            raise PipelineConfigError(f"Cannot add '{name}': it will create a loop in the pipeline.")
+        self.graph, self.root_node = _add_node_to_pipeline_graph(
+            graph=self.graph,
+            root_node_name=self.root_node,
+            components=component_definitions,
+            node={"name": name, "inputs": inputs},
+            instance=component,
+        )
```

Comment:
I had the same thought on this one: having the method returning the root node is quite confusing. I will see if there's an option to not return that value: maybe I can use another method do retrieve such node in a separate call. Something like:
```python
self.graph = _add_node_to_pipeline_graph(
            graph=self.graph,
            root_node_name=self.root_node,
            components=component_definitions,
            node={"name": name, "inputs": inputs},
            instance=component,
        )
self.root_node = _get_root_node(graph=self.graph)
```
or something along these lines. Do you think it would be better?

Reply:
I actually got a better idea. I made `self.root_node` a property, so as soon as you set the new graph, the root node will be always fetched properly without the need to set it explicitly.

Reply:
That's a very good solution. I like it.